pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java

```diff
@@ -227,70 +230,98 @@ private boolean isDone() {
-              _statusDescription = String
-                  .format("partition=%s, expected=%s, found=%s, %s", partitionName,
-                      idealStateStatus, currentStateStatus, descriptionSuffix);
-              return Status.STARTING;
-            }
-          }
+      T helixState = getState(resourceName);
+      if (helixState == null) {
+        String description = String.format("Helix state does not exist: %s", descriptionSuffix);
+        return new Pair(Status.STARTING, description);
+      }
+
+      // Check that all partitions that are supposed to be in any state other than OFFLINE have the same status in the
+      // external view or went to ERROR state (which means that we tried to load the segments/resources but failed for
+      // some reason)
+      Map<String, String> partitionStateMap = getPartitionStateMap(helixState);
+      for (String partitionName : idealState.getPartitionSet()) {
+        String idealStateStatus = idealState.getInstanceStateMap(partitionName).get(_instanceName);
+
+        // Skip this partition if it is not assigned to this instance or if the instance should be offline
+        if (idealStateStatus == null || "OFFLINE".equals(idealStateStatus)) {
+          continue;
         }
 
-        // Resource is done starting up, remove it from the set
-        _resourceIterator.remove();
+        // If the instance state is not ERROR and is not the same as what's expected from the ideal state, then it
+        // hasn't finished starting up
+        String currentStateStatus = partitionStateMap.get(partitionName);
+        if (!idealStateStatus.equals(currentStateStatus)) {
+          if ("ERROR".equals(currentStateStatus)) {
+            LOGGER.error(String.format("Resource: %s, partition: %s is in ERROR state", resourceName, partitionName));
+          } else {
+            String description = String
+                .format("partition=%s, expected=%s, found=%s, %s", partitionName,
+                    idealStateStatus, currentStateStatus, descriptionSuffix);
+            return new Pair(Status.STARTING, description);
+          }
+        }
       }
+      return new Pair(Status.GOOD, STATUS_DESCRIPTION_NONE);
+    }
 
-      if (_resourcesToMonitor.isEmpty()) {
-        _statusDescription = STATUS_DESCRIPTION_NONE;
-        LOGGER.info("Instance {} has finished starting up", _instanceName);
-      } else {
-        _statusDescription = String.format("waitingFor=%s, numResourcesLeft=%d, numTotalResources=%d, minStartCount=%d",
-            getMatchName(), _resourcesToMonitor.size(), _numTotalResourcesToMonitor, _minResourcesStartCount);
-        LOGGER.info("Instance {} returning GOOD because {}", _statusDescription);
+    private String getResourceListAsString() {
+      if (_resourcesToMonitor.size() <= MAX_RESOURCE_NAMES_TO_LOG) {
+        return _resourcesToMonitor.toString();
       }
-
-      return Status.GOOD;
+      return "[" + _resourcesToMonitor.iterator().next() + ",...]";
```

Comment:
I am not clear about this method - why is this moving the resource iterator? It sounds its mostly a toString that restricts it to less than 5 resources logged...

Reply:
It prints the name of one table the first one and then ...
If the number of tavles are small, it just returns the entire list in string form.